src/planner/binder/statement/bind_insert.cpp

```diff
@@ -25,6 +37,261 @@ static void CheckInsertColumnCountMismatch(int64_t expected_columns, int64_t res
 	}
 }
 
+unique_ptr<ParsedExpression> ExpandDefaultExpression(const ColumnDefinition &column) {
+	if (column.DefaultValue()) {
+		return column.DefaultValue()->Copy();
+	} else {
+		return make_unique<ConstantExpression>(Value(column.Type()));
+	}
+}
+
+void ReplaceDefaultExpression(unique_ptr<ParsedExpression> &expr, const ColumnDefinition &column) {
+	D_ASSERT(expr->type == ExpressionType::VALUE_DEFAULT);
+	expr = ExpandDefaultExpression(column);
+}
+
+void QualifyColumnReferences(ParsedExpression &expr, const string &table_name) {
+	// To avoid ambiguity with 'excluded', we explicitly qualify all column references
+	if (expr.type == ExpressionType::COLUMN_REF) {
+		auto &column_ref = (ColumnRefExpression &)expr;
+		if (column_ref.IsQualified()) {
+			return;
+		}
+		column_ref.SetTableName(table_name);
```

Comment:
Ah ColumnRefExpressions are similar to Values in that they are supposed to be read-only after creation?
Then I guess we want to replace it at the unique_ptr level 👍

Reply:
I removed the SetTableName function, I also changed some internal functions of ColumnRefExpression.
To use this enum
```c++
enum class ColumnQualification : uint8_t { NAME = 0, TABLE = 1, SCHEMA = 2, CATALOG = 3, QUALIFICATION_ENUM_SIZE };
```

I kept all the existing functions in tact, and added a couple extra ones centered around getting information from the internal `column_names`, let me know if you think this is a good idea or not.

I figured with this change we can turn `column_names` into a private variable, forcing you to make a new ColumnRefExpression if you want to alter it

Reply:
Actually, this doesn't work

Because we can have structs, these assumptions are not guaranteed.
But that makes me think `IsQualified` never worked in the first place, because `struct.field` is not qualified, but the old version of IsQualified only checks for `column_names.size() > 1`

Reply:
This is the old version, with added comments for clarity
```c++
const string &ColumnRefExpression::GetTableName() const {
	// We assert that the reference is at least qualified (table.name)
	// We also assert that this qualification does not exceed the max
	D_ASSERT(column_names.size() >= 2 && column_names.size() <= 4);
	if (column_names.size() == 4) {
		// The column is assumed to be fully qualified with 'catalog.schema.table.name'
		// so we return the 'table' part
		return column_names[2];
	}
	if (column_names.size() == 3) {
		// The column is assumed to be partially qualified with 'schema.table.name'
		// so we return the 'table' part
		return column_names[1];
	}
	// The column is assumed to be qualified 'table.name'
	// so the first part is the 'table' part
	return column_names[0];
}
```

But these assumptions are wrong when we have even one level of struct qualification
(struct.field, table.struct.field, schema.table.struct.field, catalog.schema.table.struct.field)

Reply:
Yes, qualification on column names is complicated when structs are involved because the ambiguity needs to be resolved during binding.`IsQualified` and `GetTableName ` are still useful because in many places of the code you cannot refer to structs, or you only care if there is any dot at all. For example, when doing an `UPDATE ... SET` we can only refer to full columns, not partial columns.

We might still want to have a go through where these are used to see if they still make sense. For example, currently we only support referencing an alias if there is no dot at all, e.g. this works:

```sql
SELECT 42 a, a;
```

But this does not work:

```sql
SELECT {'b': 42} a, a.b;
```